src/firebolt/utils/util.py

```diff
@@ -72,6 +73,52 @@ def fix_url_schema(url: str) -> str:
     return url if url.startswith("http") else f"https://{url}"
 
 
+def get_auth_endpoint(api_endpoint: URL) -> URL:
+    """Create auth endpoint from api endpoint.
+
+    Args:
+        api_endpoint (URL): provided API endpoint
+
+    Returns:
+        URL: authentication endpoint
+    """
+    return api_endpoint.copy_with(
+        host=".".join(["id"] + api_endpoint.host.split(".")[1:])
+    )
+
+
+@contextmanager
+def nested_loop() -> Generator:
+    from trio._core._run import GLOBAL_RUN_CONTEXT  # type: ignore
+
+    s = object()
+    task, runner, _dict = s, s, s
+    if hasattr(GLOBAL_RUN_CONTEXT, "__dict__"):
+        _dict = GLOBAL_RUN_CONTEXT.__dict__
+    if hasattr(GLOBAL_RUN_CONTEXT, "task"):
+        task = GLOBAL_RUN_CONTEXT.task
+        del GLOBAL_RUN_CONTEXT.task
+    if hasattr(GLOBAL_RUN_CONTEXT, "runner"):
+        runner = GLOBAL_RUN_CONTEXT.runner
+        del GLOBAL_RUN_CONTEXT.runner
+
```

Comment:
This is a code piece from StackOverflow, which has worked for me for the nested async loop case. Not sure if we'll need this anymore since the sync/async split. I will check and remove if not